File: DevExpress-Demo/DevExpress.ProductsDemo.Wpf/VB/DevExpress.ProductsDemo.Wpf/ModulesViewModels/GridContactsModuleViewModel.vb
Imports System
Imports System.Collections.Generic
Imports System.Collections.ObjectModel
Imports System.Linq
Imports DevExpress.Mvvm
Imports DevExpress.Utils
Imports DevExpress.XtraGrid
Imports ProductsDemo.Controls

Namespace ProductsDemo.Modules

    Public Class GridContactsModuleViewModel
        Inherits GridViewModelBase(Of Contact)

        Public Sub New()
            CheckedItemType = "Card"
            InitializeIndexBarSource()
            SelectedIndexBarString = IndexBarSource(0)
        End Sub

        Public Overridable Property FilteredItemsSource As ObservableCollection(Of Contact)

        Public Overridable Property IndexBarSource As List(Of String)

        Public Overridable Property SelectedIndexBarString As String

        Public Sub SetTableView()
            ChangeView(ViewType.TableView)
        End Sub

        Public Sub SetCardView()
            ChangeView(ViewType.CardView)
        End Sub

        Public Sub GroupByAlphabet()
            GroupBy("Name.FullName")
        End Sub

        Public Sub GroupByState()
            GroupBy("Address.State")
        End Sub

        Protected Overridable ReadOnly Property NotificationService As INotificationService
            Get
                Return Nothing
            End Get
        End Property

        Protected Overridable Sub OnSelectedIndexBarStringChanged()
            UpdateFilteredItemsSource()
        End Sub

        Public Overrides Sub ShowNewItemWindow()
            ShowNewContactWindow()
        End Sub

        Private Sub InitializeIndexBarSource()
            Dim source As List(Of String) = New List(Of String)()
            For Each contact As Contact In ItemsSource
                Dim firstChar As String = contact.LastName.Substring(0, 1).ToUpper()
                If source.Contains(firstChar) Then Continue For
                source.Add(firstChar)
            Next

            source.Sort()
            source.Insert(0, "ALL")
            IndexBarSource = source
        End Sub

        Private Sub UpdateFilteredItemsSource()
            Dim source As ObservableCollection(Of Contact) = New ObservableCollection(Of Contact)()
            For Each contact As Contact In ItemsSource
                If Equals(SelectedIndexBarString, IndexBarSource(0)) OrElse (Not String.IsNullOrEmpty(contact.LastName) AndAlso contact.LastName.ToUpper().StartsWith(SelectedIndexBarString)) Then source.Add(contact)
            Next

            FilteredItemsSource = source
        End Sub

        Protected Overrides Sub OnItemsSourceCollectionChanged(ByVal sender As Object, ByVal e As Collections.Specialized.NotifyCollectionChangedEventArgs)
            MyBase.OnItemsSourceCollectionChanged(sender, e)
            UpdateFilteredItemsSource()
        End Sub

        Protected Overrides Function GetColumns() As List(Of GridColumnModel)
            Dim columns As List(Of GridColumnModel) = New List(Of GridColumnModel)()
            columns.Add(New GridColumnModel() With {.Name = "Prefix", .DisplayName = String.Empty, .AllowEditing = DefaultBoolean.False, .Width = 40, .FixedWidth = True})
            columns.Add(New GridColumnModel() With {.Name = "Name.FullName", .DisplayName = "Name", .AllowEditing = DefaultBoolean.False, .GroupInterval = ColumnGroupInterval.Alphabetical})
            columns.Add(New GridColumnModel() With {.Name = "Email", .AllowEditing = DefaultBoolean.False, .Width = 182, .FixedWidth = True})
            columns.Add(New GridColumnModel() With {.Name = "Address.State", .DisplayName = "State", .AllowEditing = DefaultBoolean.False, .Width = 46, .FixedWidth = True})
            columns.Add(New GridColumnModel() With {.Name = "Address.City", .DisplayName = "City", .AllowEditing = DefaultBoolean.False, .Width = 80, .FixedWidth = True})
            columns.Add(New GridColumnModel() With {.Name = "Phone", .AllowEditing = DefaultBoolean.False, .Mask = "(\(\d\d\d\)) \d{1,3}-\d\d\d\d", .Width = 90, .FixedWidth = True})
            columns.Add(New GridColumnModel() With {.Name = "Activity", .AllowEditing = DefaultBoolean.False, .Width = 70, .FixedWidth = True})
            Return columns
        End Function

        Protected Overrides Sub InitializeDefaultView()
            ChangeView(ViewType.CardView)
            CheckedItemType = ItemType.Card
        End Sub

        Protected Overrides Function GetItemsSource() As List(Of Contact)
            Return DataBaseHelper.Instance.Contacts
        End Function

        Protected Overrides Sub GroupBy(ByVal fieldName As String)
            ChangeView(ViewType.TableView)
            MyBase.GroupBy(fieldName)
        End Sub

        Public Overridable Property EditableContact As Contact

        Public Overridable Property CurrentView As ViewType

        Public Property States As List(Of String)

        Public Property Cities As List(Of String)

        Private contactWindow As EditContactWindow = Nothing

        Private Sub ChangeView(ByVal viewType As ViewType)
            ClearModel()
            ChangeViewCore(viewType)
        End Sub

        Private Sub ChangeViewCore(ByVal viewType As ViewType)
            CurrentView = viewType
        End Sub

        Private Sub ShowEditContactWindowCore(ByVal contact As Contact)
            contactWindow = New EditContactWindow()
            EditableContact = contact
            States = GetStates()
            Cities = GetCities()
            contactWindow.DataContext = Me
            contactWindow.ShowDialog()
        End Sub

        Private isNew As Boolean

        Public Sub ShowEditContactWindow()
            If SelectedItem Is Nothing Then Return
            isNew = False
            ShowEditContactWindowCore(CType(SelectedItem.Clone(), Contact))
        End Sub

        Public Sub ShowNewContactWindow()
            isNew = True
            ShowEditContactWindowCore(New Contact())
        End Sub

        Public Sub SaveContact()
            If isNew Then
                ItemsSource.Add(EditableContact)
                SelectedItem = EditableContact
                NotificationService.CreatePredefinedNotification("Contact Created", EditableContact.Name.FullName, "", EditableContact.Photo).ShowAsync()
            Else
                SelectedItem.Assign(EditableContact)
                NotificationService.CreatePredefinedNotification("Contact Changed", EditableContact.Name.FullName, "", EditableContact.Photo).ShowAsync()
            End If

            contactWindow.Close()
        End Sub

        Public Sub CloseContactWindow()
            contactWindow.Close()
        End Sub

        Public Sub DeleteContact()
            NotificationService.CreatePredefinedNotification("Contact Deleted", SelectedItem.Name.FullName, "", SelectedItem.Photo).ShowAsync()
            ItemsSource.Remove(SelectedItem)
        End Sub

        Private Function GetCities() As List(Of String)
            Dim cities As Collections.IEnumerable =(From contact In ItemsSource Select contact.Address.City).OrderBy(Function(s) s).Distinct()
            Return cities.Cast(Of String)().ToList()
        End Function

        Private Function GetStates() As List(Of String)
            Dim states As Collections.IEnumerable =(From contact In ItemsSource Select contact.Address.State).OrderBy(Function(s) s).Distinct()
            Return states.Cast(Of String)().ToList()
        End Function
    End Class

    Public Enum ViewType
        TableView
        CardView
    End Enum
End Namespace

File: DevExpress-Demo/DevExpress.ProductsDemo.Wpf/CS/DevExpress.ProductsDemo.Wpf/ModulesViewModels/GridContactsModuleViewModel.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;
using DevExpress.Mvvm;
using DevExpress.Utils;
using DevExpress.XtraGrid;
using ProductsDemo.Controls;

namespace ProductsDemo.Modules {
    public class GridContactsModuleViewModel : GridViewModelBase<Contact> {
        public GridContactsModuleViewModel() {
            CheckedItemType = "Card";
            InitializeIndexBarSource();
            SelectedIndexBarString = IndexBarSource[0];
        }
        public virtual ObservableCollection<Contact> FilteredItemsSource { get; set; }
        public virtual List<string> IndexBarSource { get; set; }
        public virtual string SelectedIndexBarString { get; set; }
        public void SetTableView() { ChangeView(ViewType.TableView); }
        public void SetCardView() { ChangeView(ViewType.CardView); }
        public void GroupByAlphabet() { GroupBy("Name.FullName"); }
        public void GroupByState() { GroupBy("Address.State"); }
        protected virtual INotificationService NotificationService { get { return null; } }
        protected virtual void OnSelectedIndexBarStringChanged() {
            UpdateFilteredItemsSource();
        }
        public override void ShowNewItemWindow() { ShowNewContactWindow(); }
        void InitializeIndexBarSource() {
            List<string> source = new List<string>();
            foreach(Contact contact in ItemsSource) {
                string firstChar = contact.LastName.Substring(0, 1).ToUpper();
                if(source.Contains(firstChar)) continue;
                source.Add(firstChar);
            }
            source.Sort();
            source.Insert(0, "ALL");
            IndexBarSource = source;
        }
        void UpdateFilteredItemsSource() {
            ObservableCollection<Contact> source = new ObservableCollection<Contact>();
            foreach(Contact contact in ItemsSource) {
                if(SelectedIndexBarString == IndexBarSource[0] || (!string.IsNullOrEmpty(contact.LastName) && contact.LastName.ToUpper().StartsWith(SelectedIndexBarString)))
                    source.Add(contact);
            }
            FilteredItemsSource = source;
        }
        protected override void OnItemsSourceCollectionChanged(object sender, System.Collections.Specialized.NotifyCollectionChangedEventArgs e) {
            base.OnItemsSourceCollectionChanged(sender, e);
            UpdateFilteredItemsSource();
        }

        protected override List<GridColumnModel> GetColumns() {
            List<GridColumnModel> columns = new List<GridColumnModel>();
            columns.Add(new GridColumnModel() { Name = "Prefix", DisplayName = string.Empty, AllowEditing = DefaultBoolean.False, Width = 40, FixedWidth = true });
            columns.Add(new GridColumnModel() { Name = "Name.FullName", DisplayName = "Name", AllowEditing = DefaultBoolean.False, GroupInterval = ColumnGroupInterval.Alphabetical });
            columns.Add(new GridColumnModel() { Name = "Email", AllowEditing = DefaultBoolean.False, Width = 182, FixedWidth = true });
            columns.Add(new GridColumnModel() { Name = "Address.State", DisplayName = "State", AllowEditing = DefaultBoolean.False, Width = 46, FixedWidth = true });
            columns.Add(new GridColumnModel() { Name = "Address.City", DisplayName = "City", AllowEditing = DefaultBoolean.False, Width = 80, FixedWidth = true });
            columns.Add(new GridColumnModel() { Name = "Phone", AllowEditing = DefaultBoolean.False, Mask = @"(\(\d\d\d\)) \d{1,3}-\d\d\d\d", Width = 90, FixedWidth = true });
            columns.Add(new GridColumnModel() { Name = "Activity", AllowEditing = DefaultBoolean.False, Width = 70, FixedWidth = true });
            return columns;
        }

        protected override void InitializeDefaultView() {
            ChangeView(ViewType.CardView);
            CheckedItemType = ItemType.Card;
        }

        protected override List<Contact> GetItemsSource() {
            return DataBaseHelper.Instance.Contacts;
        }
        protected override void GroupBy(string fieldName) {
            ChangeView(ViewType.TableView);
            base.GroupBy(fieldName);
        }

        public virtual Contact EditableContact { get; set; }
        public virtual ViewType CurrentView { get; set; }
        public List<string> States { get; set; }
        public List<string> Cities { get; set; }

        EditContactWindow contactWindow = null;

        void ChangeView(ViewType viewType) {
            ClearModel();
            ChangeViewCore(viewType);
        }
        void ChangeViewCore(ViewType viewType) {
            CurrentView = viewType;
        }

        void ShowEditContactWindowCore(Contact contact) {
            contactWindow = new EditContactWindow();
            EditableContact = contact;
            States = GetStates();
            Cities = GetCities();
            contactWindow.DataContext = this;
            contactWindow.ShowDialog();
        }
        bool isNew;
        public void ShowEditContactWindow() {
            if(SelectedItem == null)
                return;
            isNew = false;
            ShowEditContactWindowCore((Contact)SelectedItem.Clone());
        }
        public void ShowNewContactWindow() {
            isNew = true;
            ShowEditContactWindowCore(new Contact());
        }
        public void SaveContact() {
            if(isNew) {
                ItemsSource.Add(EditableContact);
                SelectedItem = EditableContact;
                NotificationService.CreatePredefinedNotification("Contact Created", EditableContact.Name.FullName, "", EditableContact.Photo).ShowAsync();
            } else {
                SelectedItem.Assign(EditableContact);
                NotificationService.CreatePredefinedNotification("Contact Changed", EditableContact.Name.FullName, "", EditableContact.Photo).ShowAsync();
            }
            contactWindow.Close();
        }
        public void CloseContactWindow() {
            contactWindow.Close();
        }
        public void DeleteContact() {
            NotificationService.CreatePredefinedNotification("Contact Deleted", SelectedItem.Name.FullName, "", SelectedItem.Photo).ShowAsync();
            ItemsSource.Remove(SelectedItem);
        }

        List<string> GetCities() {
            System.Collections.IEnumerable cities = (from contact in ItemsSource select contact.Address.City).OrderBy(s => s).Distinct();
            return cities.Cast<string>().ToList();
        }
        List<string> GetStates() {
            System.Collections.IEnumerable states = (from contact in ItemsSource select contact.Address.State).OrderBy(s => s).Distinct();
            return states.Cast<string>().ToList();
        }
    }
    public enum ViewType { TableView, CardView }
}
